src/mpl/src/gpu/mpl_gpu_ze.c

```diff
@@ -1744,45 +1794,16 @@ int MPL_gpu_ipc_handle_create(const void *ptr, MPL_gpu_device_attr * ptr_attr,
 /* ptr must be a local device pointer and base address */
 int MPL_gpu_ipc_handle_destroy(const void *ptr, MPL_pointer_attr_t * gpu_attr)
 {
-    int status, mpl_err = MPL_SUCCESS;
-    MPL_ze_ipc_handle_entry_t *cache_entry = NULL;
+    int mpl_err = MPL_SUCCESS;
     int dev_id;
     uint64_t mem_id;
 
-    if (physical_device_states != NULL) {
-        MPL_ze_gem_hash_entry_t *entry = NULL;
-        HASH_FIND_PTR(gem_hash, &ptr, entry);
-
-        if (entry) {
-            HASH_DEL(gem_hash, entry);
-
-            /* close GEM handle */
-            for (int i = 0; i < entry->nhandles; i++) {
-                status = close_handle(physical_device_states[entry->dev_id].fd, entry->handles[i]);
-                if (status) {
-                    break;
-                }
-            }
-
-            MPL_free(entry);
-        }
-    }
-
-    if (likely(MPL_gpu_info.specialized_cache)) {
-        dev_id = device_to_dev_id(gpu_attr->device);
-        if (dev_id == -1) {
-            goto fn_fail;
-        }
-
-        mem_id = gpu_attr->device_attr.prop.id;
-        HASH_FIND(hh, ipc_cache_tracked[dev_id], &mem_id, sizeof(uint64_t), cache_entry);
-
-        if (cache_entry != NULL) {
-            free_ipc_handle_cache(cache_entry);
-            HASH_DELETE(hh, ipc_cache_tracked[dev_id], cache_entry);
-            MPL_free(cache_entry);
-        }
+    mem_id = gpu_attr->device_attr.prop.id;
+    dev_id = device_to_dev_id(gpu_attr->device);
```

Comment:
Is it safe to assume that the new allocation will be from the same device? because the `MPL_pointer_attr_t` here represents the new allocation.

Reply:
it is the property of the "ptr", which is the pointer that is being deallocated. I think this is ok.

Reply:
But the pointer we are clearing from the cache has _already been deallocated_ if the free hooks are not invoked. In which case the same pointer represents the new allocation, right?

Reply:
The pointer won't be freed until cache is cleared. So it is still valid

Reply:
Is that because we are cleaning up all the cache elements at once with this patch? Before, we could get the same pointer but with a new mem id and still find cache elements related to the old allocation. Like, was the gem hash entry not preventing the allocation from being freed?

Reply:
The fix we are trying to apply here is one where there is a hit in the gem cache for an allocation that was freed by the user, and the free hooks may not have fired. Sure, the allocation may still exist because of reference counting. What I'm trying to get at is that we probably want to store the `dev_id` of that allocation along with the `mem_id` so we don't rely on querying the _new_ pointer for the device id to know which ipc handle cache to remove it from. Where we store the dev_id I'm not entirely sure. The gem hash is not always used. So I guess the mem id hash?

Reply:
Ok, make sense. gem_hash stores the dev_id already, I will change the call to remove_stale_sender_cache to use the dev_id stored in gem_hash. However, the MPL_gpu_ipc_handle_destroy is to destroy the real (not stale) pointer before zeMemFree is called, there the dev_id is real.

Reply:
@raffenet I amended my last commit to include dev_id, please pick it from my home directory on Aurora (same directory).

Reply:
Sounds good. I think after this PR I'm going to implement a new ipc handle destroy function that takes a handle as input. We can change the existing destroy function that takes a pointer to be something like `MPL_gpu_ipc_pointer_free`.